### DISCUSSION TITLE
How do we pre-fetch the page using next/router . Tried some way but none worked.

### DISCUSSION THREAD

**Discussion by lakshmajee** [vercel/next.js#15431] [Help] [answered]
I have seen a feature available with `next/router` to pre-fetch some of the pages on the fly. (*Link* from `next/link` internally uses this). and `Link` is not a choice for my requirement.

By referring to the example available at [documentation](https://nextjs.org/docs/api-reference/next/router#routerprefetch), I've came up with following piece of code 
```javascript
try {
   await router.prefetch(
      '/[organization]/reports/[id]/view',
      `/${organization}/reports/${reportId}/view`,
   );

    // navigate to page
   router.push(`/${organization}/reports/${reportId}/view`); // <------------------- not worked
   router.push(                                           // <----------------------- not worked
       '/[organization]/reports/[id]/view',
       `/${organization}/reports/${reportId}/view`
   );
} catch (err) {
   console.log('Pre-fetch error ', err);
}
``` 
In the above snippet neither of them worked!!  `router.push(path, as)` & `router.push(path)`  .

Am i doing anthing wrong

**Comment by jamesmosier** [comment, marked as answer]
Hi @ananth-minions. 

Aside from that, in your example the second `.push` attempt *should've* worked:

```js
   router.push(
       '/[organization]/reports/[id]/view',
       `/${organization}/reports/${reportId}/view`
   );
```

when you say they didn't work, what didn't work about them? Did the page not transition? Also do you happen to have a repository with some of the code you could share?

**Reply by lakshmajee** [reply]
When i see 
```javascript
 // node_modules/next/dist/next-server/lib/router/router.d.ts  
 prefetch(url: string, asPath?: string, options?: PrefetchOptions): Promise<void>;

```
 I thought it would resolve as a promise, and asking the next line (router.push()) to wait till the previous one (router.prefetch()) resolves/completes.

When I use `router.push()` without any pre-fetch related code (router.prefetch / Link), it does the job by redirecting to the page. But it reloads the entire page i.e,  fetching entire page(s) again.

PS: ignore spell checks :D

**Reply by lakshmajee** [reply]
Does the following snippet is allow the app to reload entire page (without `as` argument)  ?
```javascript
router.push(`/${organization}/reports/${reportId}/view`);
```

**Reply by jamesmosier** [reply]
Sorry, it looks like it [does return a promise](https://github.com/jamesmosier/next.js/blob/54d991e6420d5493242a933037a8192ade0b53b8/packages/next/next-server/lib/router/router.ts#L819-L847) I was looking at the wrong `prefetch` method. 

But yes, omitting the `as` will cause a full page refresh. But `prefetch` only works in production so that is why you are seeing oddities. 

You'll see that there is a check for `NODE_ENV !== 'production'` below and that just returns, it does not resolve the promise

https://github.com/vercel/next.js/blob/e4e3ad38fd7cfc7dd5f72efd493d75a5b945c300/packages/next/next-server/lib/router/router.ts#L817-L820

**Reply by lakshmajee** [reply]
Oh! yes, you're right!  Ignoring `as` causes entire page to reload!. Suppling `as` to `push` solved my issue.!!

Thanks @jamesmosier

**Reply by lakshmajee** [reply]
Sorry reopening this issue, as I've got an alternative solution with `as`.
But what if we want to programmatically pre-fetch the page with `router.prefetch` and then trigger a redirect with `router.push()`? (actually, the original question is this, but in the middle I've got the solution to m current problem)

**Reply by jamesmosier** [reply]
You could check for the NODE_ENV before performing the prefetch

```js
if (process.env.NODE_ENV === 'production') {
  // only resolves in production
  await router.prefetch(
    '/[organization]/reports/[id]/view',
    `/${organization}/reports/${reportId}/view`
  );
}

router.push(
  '/[organization]/reports/[id]/view',
  `/${organization}/reports/${reportId}/view`
);
```

**Reply by lakshmajee** [reply]
Oh! I missed that previous comment !!
Marked as the answer:+1:
